larpandoradlcontent/LArTwoDReco/DLThreeDClusterSplittingAlgorithm.cc

```diff
@@ -0,0 +1,1350 @@
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::EndFragmentation(*this, clusterListToSaveName, clusterListToDeleteName));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+// Training Functions
+//------------------------------------------------------------------------------------------------------------------------------------------
+#ifdef MONITORING
+
+void DLThreeDClusterSplittingAlgorithm::FillHitListMaps(const Cluster *const pCluster, MCParticleToHitListMap &mainHitListMap, 
+    MCParticleToHitListMap &contHitListMap) 
+{
+    CaloHitList clusterHits;
+    LArClusterHelper::GetAllHits(pCluster, clusterHits);
+
+    for (const CaloHit *const pCaloHit : clusterHits)
+    {
+        // Identify contributing MCParticles
+        const MCParticleWeightMap hitMCParticleWeightMap(pCaloHit->GetMCParticleWeightMap());
+        for (const auto &entry : hitMCParticleWeightMap)
+        {
+            if (entry.second > m_contFraction)
+                contHitListMap[entry.first].push_back(pCaloHit);
+        }
+        // Identify dominant MCParticles
+        try
+        {
+            const MCParticle *const pMCParticle(MCParticleHelper::GetMainMCParticle(pCaloHit));
+            mainHitListMap[pMCParticle].push_back(pCaloHit);
+        }
+        catch (...) { continue; }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void DLThreeDClusterSplittingAlgorithm::FindContaminants(const MCParticleToHitListMap &mainHitListMap, const MCParticleToHitListMap &contHitListMap,
+    MCContaminantVector &mcContaminants)
+{
+    // Sort
+    MCParticleVector mcParticleVector;
+    for (const auto &entry : mainHitListMap) mcParticleVector.push_back(entry.first);
+    std::sort(mcParticleVector.begin(), mcParticleVector.end(), PointerLessThan<MCParticle>());
+
+    // Find contaminants
+    for (const MCParticle *const pMCContaminant : mcParticleVector)
+    {
+        // Get hits MCParticle contributed to
+        if (contHitListMap.find(pMCContaminant) == contHitListMap.end())
+            continue;
+
+        const CaloHitList &mainHitList(mainHitListMap.at(pMCContaminant));
+        const CaloHitList &contHitList(contHitListMap.at(pMCContaminant));
+
+        // Can we clearly see the particle in the cluster?
+        unsigned int hitCount(0);
+        for (const CaloHit *const pCaloHit : mainHitList)
+        {
+            const MCParticleWeightMap hitMCParticleWeightMap(pCaloHit->GetMCParticleWeightMap());
+            for (const auto &[pMCParticle, weight] : hitMCParticleWeightMap)
+            {
+                if (weight > m_mainFraction)
+                {
+                    ++hitCount;
+                    break;
+                }
+            }
+        }
+
+        if (hitCount < m_minTargetMCHits)
```

Comment:
Immediate reaction is I'm a bit surprised to see sub "threshold" MC being skipped before the part that builds the contaminant list. Is this because some contributions are so small that you don't even want them to be considered as contaminants?